lib/plugins/aws/package/compile/functions/index.js

```diff
@@ -441,98 +441,125 @@ class AwsCompileFunctions {
         fileHash.setEncoding('base64');
         versionHash.setEncoding('base64');
 
-        // Read the file in chunks and add them to the hash (saves memory and performance)
-        versionCompilationPromise = BbPromise.fromCallback(cb => {
-          const readStream = fs.createReadStream(artifactFilePath);
-
-          readStream
-            .on('data', chunk => {
-              fileHash.write(chunk);
-              versionHash.write(chunk);
-            })
-            .on('close', () => {
-              cb();
-            })
-            .on('error', error => {
-              cb(error);
+        // Include function and layer configuration in version id hash (without the Code part)
+        versionCompilationPromise = this.addFileContentsToHashes(artifactFilePath, [
+          fileHash,
+          versionHash,
+        ])
+          .then(() => {
+            // Include all referenced layer code in the version id hash
+            const functionProperties = _.omit(_.get(functionResource, 'Properties', {}), 'Code');
+            const layerProperties = this.extractLayerPropertiesFromFunction(
+              functionProperties,
+              cfTemplate
+            );
+
+            const layerZipFilePaths = [];
+            layerProperties.forEach(layer => {
+              const layerZipFileName = layer.properties.Properties.Content.S3Key.split(
+                path.sep
+              ).pop();
+              const layerZipFilePath = path.join(this.packagePath, layerZipFileName);
+              layerZipFilePaths.push(layerZipFilePath);
```

Comment:
This looks as not necessarily equivalent to: https://github.com/serverless/serverless/blob/01a03d103bde0476690ab2e61c81a9f00ae607df/lib/plugins/aws/package/compile/layers/index.js#L31-L34

I believe safest way, would be to resolve an actual layer configuration object for given layer, and follow above logic to resolve path to layer artifact.

Reply:
This doesn't seem to be addressed

Reply:
I looked into addressing this but an issue is that in the linked logic it uses `layerObject` which is defined using:

```
    const layerObject = this.serverless.service.getLayer(layerName);
```
where `layerName` is actually the layer key used in the serverless.yml file, not the specific layer name field.

So the challenge would be how to get the serverless.yml layer key from the Ref key provided under the function layer, i.e. how to get `testLayer` from `{ Ref: TestLayerLambdaLayer }` and also `TestLayer` from `{ Ref: TestLayerLambdaLayer}` as the key is capitalized. Any ideas? Obviously could just split by `LambdaLayer` then if an error is thrown try with lowercase first letter, but it wouldn't be very clean.

Reply:
I actually tried the split/lowercase method and it doesn't seem too bad, but let me know what you think: https://github.com/serverless/serverless/pull/8066/commits/a6ec21d1a51fd4c7674c31adacf835a348f44ab6#diff-771a8504e96d8db95f70f2cca6e3a906R458

Reply:
Probably most bulletproof way, would be to introduce a hidden property on layer resource, that references `layerObject`

We can do following

```javascript
Object.defineProperty(newLayer, "_config", { value: layerObject });
```

At https://github.com/serverless/serverless/blob/5f85543d8e47a9afaec6811a57e1235e82949550/lib/plugins/aws/package/compile/layers/index.js#L28

It's safe for CF template handling, as hidden properties don't go down to JSON, and in our logic we can access the config as:

```
cicfTemplate.Resources[potentialLocalLayerObject.Ref]._config
```

Reply:
I'm not sure if this is possible as it looks like most resources are added using `_.merge(this.serverless.service.provider.compiledCloudFormationTemplate...` which will drop the hidden (`enumerable: false`) value. If `enumerable: true` then it seems to go down to JSON.

Is there somewhere else on the main serverless object it could be stored? Could add a method to `lib/classes/Service.js` but not sure that would make it much better. I wonder if there was an "official" way of resolving between the two then this https://github.com/serverless/serverless/blob/master/lib/plugins/aws/package/compile/functions/index.js#L449 could support just a list of layer object names from serverless.yml, rather than having to use the `{ Ref: MyLayerLambdaLayer}` syntax for layers within the same stack.

Reply:
> _.merge(this.serverless.service.provider.compiledCloudFormationTemplate... which will drop the hidden (enumerable: false) 

We may replace `_.merge` with `Object.assign`. Note `_.merge` is used in many places without a valid reasons, and as it's a deep merge occasionally it causes an unexpected issues.

Reply:
This is now changed here: https://github.com/serverless/serverless/pull/8066/commits/231cf14c285086836374bca74a3300aaa7d147ca